src/Database/Esqueleto/Internal/Internal.hs

```diff
@@ -3268,6 +3318,12 @@ makeLocking info (PostgresLockingClauses clauses) =
             plain v = (v,[])
 makeLocking _ NoLockingClause = mempty
 
+makeReturning :: SqlSelect a r
+              => IdentInfo -> ReturningClause -> a -> (TLB.Builder, [PersistValue])
+makeReturning _    ReturningNothing _ = mempty
+makeReturning info ReturningStar  ret = ("RETURNING ", []) <> sqlSelectCols info ret
```

Comment:
This is getting the `SqlSelect a r` instance and using that to provide `sqlSelectCols info ret` here.

A typical `update` in `esqueleto` has type `update :: (SqlExpr (Entity val) -> SqlQuery ()) -> SqlPersistT m ()`. This would give us an `a` of `()`, which would use this instance:

```haskell
-- | Not useful for 'select', but used for 'update' and 'delete'.
instance SqlSelect () () where
    sqlSelectCols _ _ = ("1", [])
    sqlSelectColCount _ = 1
    sqlSelectProcessRow _ = Right ()
```

`RETURNING 1`.

*If* this clause `ReturningStar` is enabled in a *non-update/insert/delete* - ie,

```haskell
oops = select do
  t <- from $ table @Foo
  tellReturning ReturningStar
  pure t
```

Then we'll generate a query like:

```sql
SELECT foo.a, foo.b, foo.c
FROM foo
RETURNING foo.a, foo.b, foo.c
```

Which is definitely an error.

So the possibility of error here makes me feel like exploring a different approach.

Reply:
Sure, I though about that, but `tellReturning` is in `Internal` module... its haddock says so, too:

```haskell
-- | (Internal) Remember a @RETURNING@ clause in a query
tellReturning :: ReturningClause -> SqlQuery ()
```

The public interface is supposed to be only the added functions in `Database.Esqueleto.PostgreSQL` (plus associated instances).

Further, the rest of query builder logic didn't seem like it provided any sql correctness guarantees internally.

Reply:
Yeah,it's not a generally safe thing, but I do want to avoid adding more unsafety where possible.